lib/krb5/pac.c

```diff
@@ -1201,3 +1306,223 @@ _krb5_pac_sign(krb5_context context,
 	krb5_storage_free(spdata);
     return ret;
 }
+
+KRB5_LIB_FUNCTION krb5_error_code KRB5_LIB_CALL
+_krb5_pac_get_kdc_checksum_info(krb5_context context,
+				krb5_pac pac,
+				krb5_cksumtype *cstype,
+				uint16_t *rodc_id)
+{
+    krb5_error_code ret;
+    krb5_storage *sp = NULL;
+    const struct PAC_INFO_BUFFER *sig;
+    size_t cksumsize, prefix;
+    uint32_t type = 0;
+
+    *cstype = 0;
+    *rodc_id = 0;
+
+    sig = pac->privsvr_checksum;
+    if (sig == NULL) {
+	krb5_set_error_message(context, EINVAL, "PAC missing kdc checksum");
+	return EINVAL;
+    }
+
+    sp = krb5_storage_from_mem((char *)pac->data.data + sig->offset_lo,
+			       sig->buffersize);
+    if (sp == NULL)
+	return krb5_enomem(context);
+
+    krb5_storage_set_flags(sp, KRB5_STORAGE_BYTEORDER_LE);
+
+    ret = krb5_ret_uint32(sp, &type);
+    if (ret) {
+	krb5_storage_free(sp);
+	return ret;
+    }
+
+    ret = krb5_checksumsize(context, type, &cksumsize);
+    if (ret) {
+	krb5_storage_free(sp);
+	return ret;
+    }
+
+    *cstype = type;
+
+    prefix = krb5_storage_seek(sp, 0, SEEK_CUR);
+
+    if (cksumsize == (sig->buffersize - prefix)) {
```

Comment:
Maybe instead of an early return, check for `cksumsize > (sig->buffersize - prefix)` and handle the RODC path in a new block (that checks for `cksumsize + 2 >= (sig->buffersize - prefix)` in case it ever gets further expanded).

Reply:
Ok,I guess you meant '(sig->buffersize - prefix) >= cksumsize + 2' though.